r2/service.go

```diff
@@ -122,6 +125,62 @@ type r2Handler struct {
 	auth  auth.HTTPAuthService
 }
 
+type routeFunc func(s *service, r *http.Request) (data interface{}, err error)
+
+type serviceMetrics struct {
+	sync.RWMutex
+	scope tally.Scope
+
+	// Endpoints mapped to metrics (endpoint = path + method + status).
+	metrics map[string]serviceMetric
+}
+
+type serviceMetric struct {
+	timer   tally.Timer
+	counter tally.Counter
+}
+
+// TODO(sandusky) Wrap each route with this function.
+func handleRoute(rf routeFunc, s *service, r *http.Request, namespace string) (interface{}, error) {
+	start := time.Now()
+	data, err := rf(s, r)
+	dur := time.Since(start)
+	s.serviceMetrics.recordMetric(r.RequestURI, r.Method, namespace, dur, err)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+func (sm *serviceMetrics) recordMetric(path, method, namespace string, d time.Duration, err error) {
+	status := "success"
+	if err != nil {
+		status = "error"
+	}
+	endpoint := path + method + status
```

Comment:
Perhaps it's better to create a struct like so:
``` 
type serviceMetricID struct {
  Path string
  Method string
  Status string
}
``` 
and use that as map key instead of the concatenated string?